keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+superseded-by:
+---
+
+# Structured Parameters with the JSON-schema
+
+## Table of Contents
+
+<!--ts-->
+   * [Structured Parameters with the JSON-schema](#structured-parameters-with-the-json-schema)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+         * [Proposal Overview](#proposal-overview)
+         * [Package Format](#package-format)
+         * [CRDs](#crds)
+            * [Conversion](#conversion)
+            * [Operator](#operator)
+            * [OperatorVersion](#operatorversion)
+            * [Instance](#instance)
+         * [Client-side](#client-side)
+            * [Instance updates from the command line (-p option)](#instance-updates-from-the-command-line--p-option)
+            * [Instance updates with a parameter file (-P option)](#instance-updates-with-a-parameter-file--p-option)
+            * [Add Parameter Wizard kudo package add parameter](#add-parameter-wizard-kudo-package-add-parameter)
+            * [Parameter Listing kudo package list parameters](#parameter-listing-kudo-package-list-parameters)
+      * [Additional attributes in JSON-schema](#additional-attributes-in-json-schema)
+         * [Trigger](#trigger)
+         * [Immutable](#immutable)
+         * [OldName](#oldname)
+      * [Operator conversion and upgrade example](#operator-conversion-and-upgrade-example)
+         * [kudo.dev/v1beta1 operator version](#kudodevv1beta1-operator-version)
+         * [Conversion to kudo.dev/v1beta2](#conversion-to-kudodevv1beta2)
+         * [Upgrading operator](#upgrading-operator)
+      * [Resources](#resources)
+
+<!--te-->
+
+## Summary
+
+Replace current flat "list of parameters" with a single, nested and typed structure described by [JSON-schema](https://json-schema.org/).
+
+## Motivation
+
+Currently, KUDO uses a flat [list of parameter](https://github.com/kudobuilder/kudo/blob/main/pkg/apis/kudo/v1beta1/parameter_types.go#L4) which, while being simple and accessible, has its limits like the need to use name prefixes to group the parameters. [JSON-schema](https://json-schema.org) is a de-facto standard when describing a data format. It can also be expressed as YAML to fit into the Kubernetes ecosystem. Along with naming and descriptions it provides following improvements over the existing parameter schema:   
+- Labels
+- Nested parameters
+- Validation and ranged values
+- More parameter types (enums, numbers, etc.)
+- Easily extendable with custom attributes like `trigger` and `immutable`
+- Rich ecosystem of tools and libraries e.g. [UI generators](http://json-schema.org/implementations.html#generators-from-schemas)
+
+Latter point might help to build an ecosystem around KUDO operators which might help driving KUDOs adoption.
+
+### Goals
+
+Introduce a new parameter format based on the JSON-schema.
+
+### Non-Goals
+
+Extend the current format to support new features.
+
```

Comment:
additionally will the grouping / hierarchy be limited? and is that a goal or non-goal?   I believe dcos params were limited to 1 level deep... so it was possible to set a grouping or category of information.   This again also has an impact on the UI team and the UX... it is really hard to support infinite nesting of properties... but it is reasonable to expect 1 or 2 levels.   One can expect that the UI team would create a simple tab for each category and list all the properties in that category... challenging to imagine what infinite nesting looks like.

Reply:
Limiting the hierarchy is not a goal at this point. This is probably better solved "by convention" anyway. Also, dcos C* has e.g. [four levels](https://github.com/mesosphere/universe/blob/version-3.x/repo/packages/C/cassandra/900/config.json#L108).